src/search/SearchMessagesScreen.js

```diff
@@ -71,37 +73,39 @@ class SearchMessagesScreen extends PureComponent<Props, State> {
   // invalidate outstanding requests on change will require more work.
 
   handleQueryChange = async (query: string) => {
-    const id = ++this.lastIdSent;
-
-    if (query === '') {
-      // The empty query can be resolved without a network call.
-      this.lastIdReceived = id;
-      this.lastIdSuccess = id;
-      this.setState({ messages: null, isFetching: false });
-      return;
-    }
-
-    this.setState({ isFetching: true });
-    try {
-      const messages = await this.fetchSearchMessages(query);
-
-      // Update `state.messages` if this is our new latest result.
-      if (id > this.lastIdSuccess) {
+    this.setState({ query }, async () => {
+      const id = ++this.lastIdSent;
+
+      if (this.state.query === '') {
```

Comment:
Changed in the new revision to update the query along with state.message since its much simpler than what I had done and I agree with your point on keeping things at sync, let me know if the way I have done it is correct.

> I don't actually see where we're ever setting state.query to a new value. Here's how this two-argument form of setState is meant to be used:

I used a different implementation of `setState`, with the signature
```javascript
setState(updater, [callback])
```
It can be found at this link https://reactjs.org/docs/react-component.html#setstate

So the first argument is where I set `state.query`. It can be an object that will be shallow merged or a function which can return an object to be shallow merged.

I think the callback runs after the update / rerender. `setState` is not allowed inside the updater but you can use it inside a callback, a warning (shown below) when you try to use `setState` inside updater function also indicates that this is ok to do.

![Screenshot from 2021-07-21 23-00-42](https://user-images.githubusercontent.com/40268170/126533699-b84cf5af-4ccb-4373-a931-b5b3840eaa51.png)

I could have used `componentDidUpdate` here as well I think.

Reply:
> ```js
> setState(updater, [callback])
> ```
>   
> It can be found at this link https://reactjs.org/docs/react-component.html#setstate

Ah I see, thanks! I think I hadn't been aware of that second argument before.

And looking back at the bit of docs I'd linked, that's not actually using the two-argument form -- it's just passing the one first argument, the "updater", and doing so with a function rather than an object.



> Changed in the new revision to update the query along with state.message since its much simpler than what I had done and I agree with your point on keeping things at sync

:+1: